### PR TITLE
feat(android): added new methods in Ti.Calendar.Calendar module for bulk operations

### PR DESCRIPTION
**Pull request by prashantsaini1**
This PR brings major improvements to deal with bulk operations in [Ti.Calendar.Calendar](https://titaniumsdk.com/api/titanium/calendar/calendar.html) module. Test results are below in this PR.

### 1. createEvents(args)
   1. **args** - an array of [Ti.Calendar.Event properties](https://titaniumsdk.com/api/titanium/calendar/event.html), same as [createEvent signature](https://titaniumsdk.com/api/titanium/calendar/calendar.html#methods_createevent) in an array.
   2. **Returns** - an array containing [Ti.Calendar.Event](https://titaniumsdk.com/api/titanium/calendar/event.html) for successfully created events, or **null** for failed events, on same position as passed in args so that the results can be properly used.

### 2. deleteEvents(ids)
   1. **ids** - an array of [Ti.Calendar.Event.id](https://titaniumsdk.com/api/titanium/calendar/event.html#properties_id) value, both int/string are accepted.
      1. If **ids** are known upfront without having to create a calendar event, same can be passed directly, thus removing the need to first create an event and then call [remove()](https://titaniumsdk.com/api/titanium/calendar/event.html#methods_remove) method on it.
   3. **Returns** - an integer value denoting how many events were deleted. If 0, either it was a failed operation, or there were actually no events associated with the passed **ids**. Can be `<= ids.length`.
      1. This is by design in native SDK which only returns the count, not the success result in delete operation. For single event delete operation, a check of `== 1` can be applied to know if it was a successful deletion or not.

### 3. getEventsById(ids)
   1. **ids** - same as in `deleteEvents()` above.
   2. **Returns** - an array containing [Ti.Calendar.Event](https://titaniumsdk.com/api/titanium/calendar/event.html) for successfully created events.
      1. The returned array length can be `<= ids.length` if some events are not created. Generally the `id` property of created events can be used to compare with the passed `ids` array to know which events got successfully created.

## Test results to show approx time taken for 100 events operations.

| - | Creation | Deletion | Fetch |
| - | - | - | - |
| createEvent() loop | 2500 ms |  |  |
| createEvents() | 500 ms |  |  |
| remove() loop | | 3800 ms  |  |
| deleteEvents() | | 22 ms  |  |
| getEventById() loop | | | 3200 ms |
| getEventsById() | | | 25 ms |

### CODE REVIEW

File: android/modules/calendar/src/java/ti/modules/titanium/calendar/CalendarProxy.java [hunk 1]
```diff
@@ -188,12 +198,147 @@ public EventProxy getEventById(int id)
 			return null;
 	}
 
+	@Kroll.method
+	public EventProxy[] getEventsById(Object args)
```
Comment (hansemannn) [inline android/modules/calendar/src/java/ti/modules/titanium/calendar/CalendarProxy.java:202]: Please add docs for the new APIs

**Review by hansemannn — CHANGES_REQUESTED** [review]
If you can add docs and an example test to test, this should work out.

**Review by m1ga — COMMENTED** [review]
just a small change as we have TiC properties for those parts

**Comment by m1ga** [comment]
will there be iOS parity? Guess it would be nice to have those methods there too

**Comment by m1ga** [comment]
Got it :+1: I didn't test it yet but sounds great that it is a performance boost!

**Comment by prashantsaini1** [comment]
@m1ga Could you please review the format of docs added in yaml if they are fine?

File: android/titanium/src/java/org/appcelerator/titanium/proxy/TiWindowProxy.java [hunk 2]
```diff
@@ -568,6 +568,9 @@ protected void fillIntent(Activity activity, Intent intent)
 			// We're opening child activity from Titanium root activity. Have it exit out of app by default.
 			// Note: If launched via startActivityForResult(), then root activity won't be the task's root.
 			intent.putExtra(TiC.INTENT_PROPERTY_FINISH_ROOT, true);
+
+			// Set default value as stated in docs on first window also if not already set above.
+			setProperty(TiC.PROPERTY_EXIT_ON_CLOSE, true);
```
Comment (m1ga) [inline android/titanium/src/java/org/appcelerator/titanium/proxy/TiWindowProxy.java:573]: are you sure this belongs into this PR?

**Comment by m1ga** [comment]
> @m1ga Could you please review the format of docs added in yaml if they are fine?

you can run `npm run lint:docs` to see the errors

File: android/titanium/src/java/org/appcelerator/titanium/proxy/TiWindowProxy.java (hunk 2 above)
Comment (prashantsaini1) [reply android/titanium/src/java/org/appcelerator/titanium/proxy/TiWindowProxy.java]: I think it slipped into this PR as it was done on the other PR we were testing out related to base activity. Thanks for the catch!

File: android/modules/calendar/src/java/ti/modules/titanium/calendar/CalendarProxy.java [hunk 3]
```diff
@@ -188,12 +198,147 @@ public EventProxy getEventById(int id)
...
+	{
+		// Validate arguments to be an array.
+		if (!(data instanceof Object[] dataList && dataList.length > 0)) {
+			Log.e(TAG, "Argument expected to be an array.");
+			return null;
+		}
+
+		// Check for permissions.
+		ContentResolver contentResolver = TiApplication.getInstance().getContentResolver();
+		if (!CalendarProxy.hasCalendarPermissions()) {
```
Comment (m1ga) [inline android/modules/calendar/src/java/ti/modules/titanium/calendar/CalendarProxy.java:233]:
```suggestion
		if (!hasCalendarPermissions()) {
```

File: android/modules/calendar/src/java/ti/modules/titanium/calendar/CalendarProxy.java [hunk 4]
```diff
@@ -188,12 +198,147 @@ public EventProxy getEventById(int id)
...
+
+		// Validate arguments to be an array.
+		if (!(args instanceof Object[] eventIds && eventIds.length > 0)) {
+			Log.e(TAG, "Argument expected to be an array.");
+			return deletedCount;
+		}
+
+		// Check for permissions.
+		ContentResolver contentResolver = TiApplication.getInstance().getContentResolver();
+		if (!CalendarProxy.hasCalendarPermissions()) {
```
Comment (m1ga) [inline android/modules/calendar/src/java/ti/modules/titanium/calendar/CalendarProxy.java:309]:
```suggestion
		if (!hasCalendarPermissions()) {
```

File: android/modules/calendar/src/java/ti/modules/titanium/calendar/CalendarProxy.java [hunk 5]
```diff
@@ -188,12 +198,147 @@ public EventProxy getEventById(int id)
...
+			KrollDict krollDict = new KrollDict((HashMap) dataList[i]);
+
+			EventProxy eventProxy = new EventProxy();
+			ContentValues contentValues = CalendarUtils.createContentValues(this, krollDict, eventProxy);
+
+			// We cannot pass null data to ContentProviderOperation.
+			// Necessary to keep track of non-null items later.
+			if (contentValues == null) {
+				eventProxies.add(null);
+				Log.e(TAG, "Title was not created, no title found for event");
```
Comment (m1ga) [inline android/modules/calendar/src/java/ti/modules/titanium/calendar/CalendarProxy.java:252]:
```suggestion
				Log.e(TAG, "Event was not created, no title found for event");
```

Might be better?

File: android/modules/calendar/src/java/ti/modules/titanium/calendar/CalendarProxy.java [hunk 6]
```diff
@@ -188,12 +198,147 @@ public EventProxy getEventById(int id)
...
+		int deletedCount = 0;
+
+		// Validate arguments to be an array.
+		if (!(args instanceof Object[] eventIds && eventIds.length > 0)) {
+			Log.e(TAG, "Argument expected to be an array.");
+			return deletedCount;
+		}
+
+		// Check for permissions.
+		ContentResolver contentResolver = TiApplication.getInstance().getContentResolver();
```
Comment (m1ga) [inline android/modules/calendar/src/java/ti/modules/titanium/calendar/CalendarProxy.java:308]: move that into the `try` part so it is closer the the place where it is used and if we don't have permissions we don't need to get it.

**Review by m1ga — COMMENTED** [review]
some minor code changes. Builds fine

File: apidoc/Titanium/UI/ListView.yml [hunk 7]
```diff
@@ -659,6 +659,30 @@ events:
...
+        summary: |
+            The index of the first visible item in the list view when the event fires; this item might not be fully visible.
+            Note: The index is `-1` when there are no items in the <Titanium.UI.ListView>.
+        type: Number
+
+      - name: firstVisibleSectionIndex
+        summary: |
+            The index of the first visible section in the list view when the event fires.
+            Note: The index is `-1` when there are no items in the <Titanium.UI.ListView>.
+        type: Number
```
Comment (m1ga) [inline apidoc/Titanium/UI/ListView.yml:685]: If those are in the scrolling event I think this should be fine as it is just a doc update :smile: But I guess it is from a different branch too :wink:

Comment (prashantsaini1) [reply apidoc/Titanium/UI/ListView.yml:685]: They were from this branch itself as they were in my checklist from few weeks, so I though it's fine to put them here also 😄

**Comment by prashantsaini1** [comment]
@m1ga This should help you test the changes:
```
const eventPayload = {
	name: "Test Event",
	startDate: "2024-12-24",
	endDate: "2024-12-25",
	location: "Titanium SDK"
};

var window = Ti.UI.createWindow();
var view = Ti.UI.createView({ layout: 'vertical', height: Ti.UI.SIZE });

var b1 = Ti.UI.createButton({
	top: 60,
	title: 'Calendar: Run Bulk Ops'
});
b1.addEventListener('click', runBulkOps);

var b2 = Ti.UI.createButton({
	top: 60,
	title: 'Calendar: Run Single Ops'
});
b2.addEventListener('click', runSingleOps);


var l1 = Ti.UI.createLabel({
	top: 60,
	left: 40,
	right: 40,
	text: 'Make sure to run this action to cleanup all test events.'
});
var b3 = Ti.UI.createButton({
	top: 20,
	title: 'Calendar: Clear All'
});
b3.addEventListener('click', clearAll);

view.add(b1);
view.add(b2);
view.add(l1);
view.add(b3);

window.add(view);

window.open();

function permissionAllowed() {
	return new Promise(resolve => {
		Ti.Calendar.requestCalendarPermissions(event => resolve(event.success));
	});
}

function storeIDs(ids) {
	// Store very first event ID only for once.
	const firstEventId = Ti.App.Properties.getInt('first_event_id', 0);
	if (firstEventId === 0) {
		Ti.App.Properties.setInt('first_event_id', ids[0]);
	}

	// Update last event id on every ops.
	Ti.App.Properties.setInt('last_event_id', ids[ids.length - 1]);
}

async function runBulkOps() {
	if (!await permissionAllowed()) {
		alert('allow calendar permissions');
		return;
	}

	const calendar = Ti.Calendar.defaultCalendar;
	const ids = createEventsInBulk(calendar);
	storeIDs(ids);
	getEventsInBulk(calendar, ids);
	deleteEventsInBulk(calendar, ids);
}

async function runSingleOps() {
	if (!await permissionAllowed()) {
		alert('allow calendar permissions');
		return;
	}
	
	const calendar = Ti.Calendar.defaultCalendar;
	const ids = createEventsOneByOne(calendar);
	storeIDs(ids);
	getEventOneByOne(calendar, ids);
	deleteEventOneByOne(calendar, ids);
}

// Single Ops.
function createEventsOneByOne(calendar) {
	const ids = [];
	for (let i = 1; i <= 10; i++) {
		const event = calendar.createEvent( createCalendarEventPayload(eventPayload, i) );
		ids.push(event.id);
	}
	return ids;
}

function getEventOneByOne(calendar, ids) {
	return ids.map(id => calendar.getEventById(id));
}

function deleteEventOneByOne(calendar, ids) {
	ids.forEach(id => calendar.getEventById(id)?.remove());
}


// Bulk Ops.
function createEventsInBulk(calendar) {
	const eventsData = [];
	for (let i = 1; i <= 10; i++) {
		eventsData.push( createCalendarEventPayload(eventPayload, i) );
	}
	
	return calendar.createEvents(eventsData).map(event => event.id);
}

function getEventsInBulk(calendar, ids) {
	return calendar.getEventsById(ids);
}

function deleteEventsInBulk(calendar, ids) {
	return calendar.deleteEvents(ids);
}

function createCalendarEventPayload(calendarItem, count) {
	return {
		title: calendarItem.name + ' : Count ' + count,
		begin: new Date(calendarItem.startDate),
		end: new Date(calendarItem.endDate),
		allDay: true,
		location: calendarItem.location
	};
}

// Clear all test events.
function clearAll() {
	const calendar = Ti.Calendar.defaultCalendar;

	const firstEventId = Ti.App.Properties.getInt('first_event_id', 0);
	const lastEventId = Ti.App.Properties.getInt('last_event_id', 0);

	const ids = Array.from({ length: lastEventId - firstEventId + 1 }, (_, i) => firstEventId + i);
	const deletedCount = deleteEventsInBulk(calendar, ids);

	if (deletedCount > 0) {
		// Reset IDs.
		Ti.App.Properties.setInt('first_event_id', 0);
		Ti.App.Properties.setInt('last_event_id', 0);
	}
}

```

**Comment by m1ga** [comment]
**Function test:**
working fine. Adding (normal and bulk) work fine and all is removed when clicking the button. I've adjusted the example code so it won't remove them right away so I could test it :)

**Comment by prashantsaini1** [comment]
@m1ga As I was doing lots of lots of testing with 100/200+ events, my device feels a hang on syncing events from Google, so I do not see events added by either way now 😄 . Sometimes the event data is null in single update methods also, sometimes it's available. So I suspect it has something to do with syncing.

**Comment by m1ga** [comment]
> @m1ga As I was doing lots of lots of testing with 100/200+ events, my device feels a hang on syncing events from Google, so I do not see events added by either way now 😄 . Sometimes the event data is null in single update methods also, sometimes it's available. So I suspect it has something to do with syncing.

yeah, I think if you delete it while it is syncing it will get that event from the cloud again. Happens sometimes when you do stuff quickly.

But it looks fine with the 10 events:

[Bildschirmaufnahme_20241204_141118.webm](https://github.com/user-attachments/assets/8b148c2f-1821-49e7-b3cf-394568e3e0c0)

**Review by hansemannn — APPROVED** [review]

**Pull request merged**